derive/src/lib.rs

```diff
@@ -42,8 +44,8 @@ pub fn from_primitive(input: TokenStream) -> TokenStream {
                     panic!("`FromPrimitive` can be applied only to unitary enums, {}::{} is either struct or tuple", name, ident)
                 },
             }
-            if let Some(val) = variant.discriminant {
-                idx = val.value;
+            if let Some(Lit(Int(value, _))) = variant.discriminant {
```

Comment:
Can we support the full `ConstExpr`, instead of picking it apart?  This will fail on even simple values like `-100` which will map to something like `Unary(Neg, Lit(Int(100)))`.  Nevermind more complicated expressions.  (It seems the old syn just failed to parse it at all.)

I'm thinking instead of `idx`, we store the actual `expr` and an `offset` 0, which we encode as `quote!(... #expr + #offset ...)`.  If the next variant has a discriminant, update `expr` and set `offset` back to 0, otherwise reuse the former `expr` with an incremented `offset`.